File: LD2.py
# IS laboratorinis darbas nr. 2. Dainius Varna EKSfm20 202000333

import math as m
import numpy as np
import random


# Sugeneruojame neurono isejime tikimasi atsakyma
def n_input(k):
    a = [0] * k
    n = 0
    m = 1 / k

    while n != len(a):
        if a[n] == 1:
            break
        else:
            a[n] = a[n-1] + m
            n += 1
    return a


# Sugeneruojame neurono isejima
def n_output(a):
    return (1 + 0.6 * m.sin(2 * m.pi * a / 0.7)) + 0.3 * (m.sin(2 * m.pi * a)) / 2


# Sigmoidinė  funkcija
def sigmoid(h):
    return 1 / 1 + np.exp(-h)


# Norimi atsakymai
def n_ats(xx):
    y_norimi = [0] * len(xx)
    a = 0
    while a != len(xx):
        y_norimi[a] = n_output(xx[a])
        a += 1

    return y_norimi


# Sigmoidine aktyvavimo funkcija
def n_aktyvimas(xx):
    y_1 = [0] * len(xx)
    a = 0

    while a != len(xx):
        y_1[a] = sigmoid(xx[a])
        a += 1

    return y_1


# Funkcija apskaiciuoti klaidas - 1/2 * (e^2)
def klaidu_tikrinimas(d, y):
    e = d - y
    return e
# m.sqrt(e[0] ** e[0])


# Funkcija w parametrams sugeneruoti pagal nurodyta kieki
def w_parametru_generavimas(n_h):
    w1 = [0] * n_h
    n = 0
    while n != len(w1):
        w1[n] = random.random()
        n += 1
    return w1


# Funkcija b parametrams sugeneruoti pagal nurodyta kieki
def b_parametru_generavimas(n_h):
    b1 = [0] * n_h
    n = 0
    while n != len(b1):
        b1[n] = random.random()
        n += 1
    return b1


# Funkcija y isejimui paskaiciuoti pagal gautus ivercius
def y1_funkcija(w_1, b_1, x_1):
    y1 = [0] * len(w_1)
    a = 0

    while a != len(w_1):
        y1[a] = x_1 * w_1[a] + b_1[a]
        a += 1

    return y1


def y_funkcija(y1, w2, b2):
    n = 0
    a = 0
    while n != len(y1):
        a = 0 + y1[n] * w2[n]
        n += 1

    return a + b2


# Funkcija tikslumui paskaiciuoti
def tikslumas(n1, n2):
    return n1 - n2


# Funkcija palyginti sugeneruota y isejima pagal formule ir algoritma
def x_mtip_compare(wt1, bt1, wt2, bt2):
    xt_n = 20  # neurono iejimu kiekis
    xtc = n_input(xt_n)
    y_1 = [0] * len(xtc)
    v_1 = [0] * len(xtc)
    y = [0] * len(xtc)  # masyvas algoritmui
    yr = [0] * len(xtc)  # masyvas atsakymam su formule (output)
    n = 0
    while n != len(xtc):
        yr[n] = n_output(xtc[n])
        n += 1
    n = 0
    while n != len(xtc):
        y_1[n] = y1_funkcija(wt1, bt1, xtc[n])
        v_1[n] = n_aktyvimas(y_1[n])
        y[n] = y_funkcija(v_1[n], wt2, bt2)
        n += 1
    print("Palyginti ats:")
    print("yr(output) =", yr)
    print("y(algoritmo) =", y)


# Funkcija isbandyti algortimo gautus ivercius
def mtip_backpropogation_test(bp_x1, bp_w1, bp_b1, bp_w2, bp_b2):
    y_1 = y1_funkcija(bp_w1, bp_b1, bp_x1)  # Issiunciame parametrus ir apskaiciuojame lygtis
    v_1 = n_aktyvimas(y_1)  # issiunciame lygties atsakymus i aktyvavimo funkcija
    y = y_funkcija(v_1, bp_w2, bp_b2)
    klaidu_tikrinimas(n_output(bp_x1), y)
    x_mtip_compare(bp_w1, bp_b1, bp_w2, bp_b2)

    return y


# Parametru atnaujinimo funkcija
def o_parametru_atnaujinimas(l_r, w2, b2, e, v1, y1, b1, w1, x_i, y):
    a = len(w2)
    v1_inv = [0] * len(w2)
    delta = [0] * len(w2)
    delta_1 = [0] * len(w2)
    n = 0
    while n != a:
        w2[n] = w2[n] + l_r * e * y
        n += 1
    b2 = b2 + l_r * e * 1

    n = 0
    while n != a:
        v1_inv[n] = v1[n] * (1 - v1[n])
        n += 1
    n = 0

    while n != a:
        delta[n] = v1_inv[n] * e * w2[n]
        n += 1
    n = 0

    while n != a-1:

        w2[n+1] = w2[n] + l_r * delta[n] * y1[n]
        w1[n+1] = w1[n] + l_r * delta[n] * x_i
        n += 1
    n = 0
    while n != a:
        delta_1[n] = (1 / 1 + m.exp((-1 * x_i) * w1[n] - b1[n]))*(1-1/(1 + m.exp((-1*x_i)*w1[n]-b1[n])))*e*w2[n]
        n += 1
    n = 0

    while n != a:
        w1[n] = w1[n] + l_r * delta_1[n] * x_i
        b1[n] = b1[n] + l_r * delta_1[n]
        n += 1
    n3 = mtip_backpropogation_test(x_i, w1, b1, w2, b2)
    # x_mtip_compare(w1, b1, w2, b2)
    # print(y_gen)
    return n3  # , w1, b1, w2, b2


# Tinklo sukurimo funkcija
def mlp_sukurimas(x1, hidden, rate):

    w1_1 = w_parametru_generavimas(hidden)  # Sugeneruojame w1 parametra pagal norima neuronu skaiciu
    b1_1 = b_parametru_generavimas(hidden)  # Sugeneruojame b1 parametra pagal norima neuronu skaiciu
    y_1 = y1_funkcija(w1_1, b1_1, x1)  # Issiunciame parametrus ir apskaiciuojame lygtis
    v_1 = n_aktyvimas(y_1)  # issiunciame lygties atsakymus i aktyvavimo funkcija
    w2_1 = w_parametru_generavimas(hidden)  # Sugeneruojame w1 parametra pagal norima neuronu skaiciu
    b2_1 = b_parametru_generavimas(1)  # Sugeneruojame b1 parametra pagal norima neuronu skaiciu
    y = y_funkcija(v_1, w2_1, b2_1)
    n1 = n_output(x1)
    e = klaidu_tikrinimas(n1, y)
    n2 = o_parametru_atnaujinimas(rate, w2_1, b2_1, e, v_1, y_1, b1_1, w1_1, x1, y)
    n3 = tikslumas(n1, n2)
    rez = 0

    # Kol skirtumas ne mazesnis jei nustatytas kartosim (0.1 / -0.1)
    while 0.1 < n3 > -0.1:
        n2 = o_parametru_atnaujinimas(rate, w2_1, b2_1, e, v_1, y_1, b1_1, w1_1, x1, y)
        n3 = tikslumas(n1, n2)
        if n3 > 2:
            rez = 1
            break
        if n3 < -2:
            rez = 1
            break
    if rez == 0:
        print("n_output:", n1)
        print("backpropogation:", n2)
        print("Skirtumas", n3)
    return rez


# Main
x_n = 20  # neurono iejimu kiekis
x = n_input(x_n)

n_target = 1  # X pavyzdzio pasirinkimas
n_hidden = 4  # pasleptu neuronu kiekis
l_rate = 0.3  # learning rate

# Kol negausime norimo atsakymo kartosime
a = 1
while a != 0:
    a = mlp_sukurimas(x[n_target], n_hidden, l_rate)

File: LD2.m
%%IS LD2 Dainius Varna EKSfm20
function LD2()
clear all;
close all;
clc;

x1_input = 0.1:1/22:1; %pirmas iejimas
y_n = (1 + 0.6*sin(2*pi*x1_input/0.7)) + 0.3*sin(2*pi*x1_input)/2; %norimas ats
n = 0.3; %mokymo zingsnis
y = backpropogation(x1_input,y_n,n); %isejimas su backpropogation

plot(x1_input,y_n,x1_input,y); grid on;
legend('Norimas ats','Backpropogation');


end
%%
%Sigmoidine aktyvavimo funkcija
function y = outS(x,w,b)
    y = 1./(1+exp(-x*w+b));
end
%%
function y = backpropogation(x,y_n,n)
    %Paruosiam parametrus:
    w_11 = [randn(1),randn(1),randn(1),randn(1),randn(1)];
    b_11 = [randn(1),randn(1),randn(1),randn(1),randn(1)];
    w_21 = [randn(1),randn(1),randn(1),randn(1),randn(1)];
    b_21 = randn(1);

    %prealocatinam del greicio
    y=zeros(1,length(x));
    e1=zeros(1,length(x));
    
    %Backpropogation
    for i=1:length(x)
        %%
        %hidden neuronu isvedimu paskaiciavimas su sigmoidine funkcija
        y1_1 = outS(x(i),w_11(1),b_11(1));
        y1_2 = outS(x(i),w_11(2),b_11(2));
        y1_3 = outS(x(i),w_11(3),b_11(3));
        y1_4 = outS(x(i),w_11(4),b_11(4));
        y1_5 = outS(x(i),w_11(5),b_11(5));
        y(i) = y1_1 * w_21(1) + y1_2 * w_21(2) + y1_3 * w_21(3) + y1_4 * w_21(4)+ y1_5 * w_21(5)+b_21;
        %%
        %apskaiciuojame  klaida:
        e1(i) = y_n(i)-y(i);
        %%
        %parametru atnaujinimas:
        w_21(1) = w_21(1) + n * e1(i) * y1_1;
        w_21(2) = w_21(2) + n * e1(i) * y1_2;
        w_21(3) = w_21(3) + n * e1(i) * y1_3;
        w_21(4) = w_21(4) + n * e1(i) * y1_4;
        w_21(5) = w_21(5) + n * e1(i) * y1_5;
        %%
        %hidden layer update:
        delta11 = (1/(1+exp(-x(i)*w_11(1)-b_11(1))))*(1/(1+exp(-x(i)*w_11(1)-b_11(1))))*e1(i)*w_21(1);
        w_11(1) = w_11(1) + n * delta11 * x(i);
        delta12 = (1/(1+exp(-x(i)*w_11(2)-b_11(2))))*(1/(1+exp(-x(i)*w_11(2)-b_11(2))))*e1(i)*w_21(2);
        w_11(2) = w_11(2) + n * delta12 * x(i);
        delta13 = (1/(1+exp(-x(i)*w_11(3)-b_11(2))))*(1/(1+exp(-x(i)*w_11(3)-b_11(3))))*e1(i)*w_21(3);
        w_11(3) = w_11(3) + n * delta13 * x(i);
        delta14 = (1/(1+exp(-x(i)*w_11(4)-b_11(4))))*(1/(1+exp(-x(i)*w_11(4)-b_11(4))))*e1(i)*w_21(4);
        w_11(4) = w_11(4) + n * delta14 * x(i);
        delta15 = (1/(1+exp(-x(i)*w_11(5)-b_11(5))))*(1/(1+exp(-x(i)*w_11(5)-b_11(5))))*e1(i)*w_21(5);
        w_11(5) = w_11(5) + n * delta15 * x(i);
        b_11(1) = b_11(1) + n * delta11;
        b_11(2) = b_11(2) + n * delta12;
        b_11(3) = b_11(3) + n * delta13;
        b_11(4) = b_11(4) + n * delta14;
        b_11(5) = b_11(5) + n * delta15;
    end
    
    
end